File: ch2/src/py_demo_pkg/py_demo_pkg/learn_thread.py
import threading
import requests

class Download:
    def __init__(self):
        pass
    
    def download(self, url, callback):
        print(f"Thread {threading.get_ident()} start downloading: {url}")
        response = requests.get(url)
        response.encoding = 'utf-8'
        callback(url, response.text)
        
    def start_download(self, url, callback):
        thread = threading.Thread(target=self.download, args=(url, callback))
        thread.start()
        

def download_finished_callback(url, result):
    print(f"{url} download finished, total nums: {len(result)}, content: {result[:5]}...")
    

def main():
    d = Download()
    d.start_download('http://localhost:8000/novel1.txt', download_finished_callback)
    d.start_download('http://localhost:8000/novel2.txt', download_finished_callback)
    d.start_download('http://localhost:8000/novel3.txt', download_finished_callback)
    
    
        
    

File: ch2/src/cpp_demo_pkg/src/learn_thread.cpp
#include <iostream>
#include <thread>
#include <chrono>
#include <functional>
#include <cpp-httplib/httplib.h>

class Download {
public:
    Download() = default;
    ~Download() = default;

    template <typename CallbackT>
    void download(const std::string& host, const std::string& path, CallbackT&& callback) {
        std::cout << "thread id: " << std::this_thread::get_id() << std::endl;
        httplib::Client client(host);
        auto response = client.Get(path);
        if (response && response->status == 200) { callback(path, response->body); }
    }

    template <typename CallbackT>
    void start_download(const std::string& host, const std::string& path, CallbackT&& callback) {
        auto thread_func = [this, host, path, callback = std::forward<CallbackT>(callback)]() {
            this->download(host, path, callback);
        };
        std::thread download_thread(std::move(thread_func));
        download_thread.detach();
    }
};

int main() {
    Download download;
    auto download_finished_callback = [](const std::string& path, const std::string& result) {
        std::cout << "Download finished! " << path << ", total: " << result.length() << std::endl;
    };

    download.start_download("http://localhost:8000", "/novel1.txt", download_finished_callback);
    download.start_download("http://localhost:8000", "/novel2.txt", download_finished_callback);
    // download.start_download("http://localhost:8000", "/novel3.txt", download_finished_callback);

    std::this_thread::sleep_for(std::chrono::milliseconds(1000 * 5));

    return 0;
}